File: app/src/main/java/co/zipperstudios/currencyexchange/utils/CurrencyExchangeDeserializer.java
package co.zipperstudios.currencyexchange.utils;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;
import co.zipperstudios.currencyexchange.data.model.CurrencyExchangeResponse;
import com.google.gson.*;

import java.lang.reflect.Type;
import java.util.HashMap;
import java.util.Map;

public class CurrencyExchangeDeserializer implements JsonDeserializer<CurrencyExchangeResponse> {

    private static final String KEY_URI = "base";
    private static final String KEY_METHOD = "date";
    private static final String KEY_PARAMETERS = "rates";

    @Override
    public CurrencyExchangeResponse deserialize(JsonElement json, Type typeOfT, JsonDeserializationContext context) throws JsonParseException {
        final JsonObject jsonObject = json.getAsJsonObject();

        // Read simple String values.
        final String uri = jsonObject.get(KEY_URI).getAsString();
        final String httpMethod = jsonObject.get(KEY_METHOD).getAsString();

        // Read the dynamic exchangeRates object.
        final Map<String, Float> parameters = readParametersMap(jsonObject);

        return new CurrencyExchangeResponse(uri, httpMethod, parameters);
    }

    @Nullable
    private Map<String, Float> readParametersMap(@NonNull final JsonObject jsonObject) {
        final JsonElement paramsElement = jsonObject.get(KEY_PARAMETERS);
        if (paramsElement == null) {
            // value not present at all, just return null
            return null;
        }

        final JsonObject parametersObject = paramsElement.getAsJsonObject();
        final Map<String, Float> parameters = new HashMap<>();
        for (Map.Entry<String, JsonElement> entry : parametersObject.entrySet()) {
            String key = entry.getKey();
            float value = entry.getValue().getAsFloat();
            parameters.put(key, value);
        }
        return parameters;
    }
}

File: app/src/main/java/co/zipperstudios/currencyexchange/utils/CurrencyExchangeDeserializer.kt
package co.zipperstudios.currencyexchange.utils

import co.zipperstudios.currencyexchange.data.model.CurrencyExchange
import co.zipperstudios.currencyexchange.data.model.CurrencyExchangeResponse
import com.google.gson.*

import java.lang.reflect.Type
import java.util.HashMap

class CurrencyExchangeDeserializer : JsonDeserializer<CurrencyExchangeResponse> {

    @Throws(JsonParseException::class)
    override fun deserialize(
        json: JsonElement,
        typeOfT: Type,
        context: JsonDeserializationContext
    ): CurrencyExchangeResponse {
        val jsonObject = json.asJsonObject

        // Read simple String values.
        val uri = jsonObject.get(KEY_BASE).asString
        val httpMethod = jsonObject.get(KEY_DATE).asString

        // Read the dynamic exchangeRates object.
        val exchangeRates = readExchangeRates(jsonObject)

        return CurrencyExchangeResponse(uri, httpMethod, exchangeRates!!)
    }

    private fun readParametersMap(jsonObject: JsonObject): Map<String, Float>? {
        val paramsElement = jsonObject.get(KEY_RATES)
            ?: // value not present at all, just return null
            return null

        val parametersObject = paramsElement.asJsonObject

        val parameters = HashMap<String, Float>()
        for ((key, value1) in parametersObject.entrySet()) {
            val value = value1.asFloat
            parameters[key] = value
        }


        return parameters
    }


    private fun readExchangeRates(jsonObject: JsonObject): HashSet<CurrencyExchange>? {
        val paramsElement = jsonObject.get(KEY_RATES)
            ?: // value not present at all, just return null
            return null

        val parametersObject = paramsElement.asJsonObject

        val exchangeRates = HashSet<CurrencyExchange>()

        for ((key, value1) in parametersObject.entrySet()) {
            exchangeRates.add(CurrencyExchange(key, value1.asFloat))
        }

        return exchangeRates
    }

    companion object {

        private val KEY_BASE = "base"
        private val KEY_DATE = "date"
        private val KEY_RATES = "rates"
    }
}
